pkg/api/kptfile/v1alpha2/types.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package defines the schema for Kptfile version v1alpha2.
+package v1alpha2
+
+import (
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const (
+	KptFileName       = "Kptfile"
+	KptFileGroup      = "kpt.dev"
+	KptFileVersion    = "v1alpha2"
+	KptFileAPIVersion = KptFileGroup + "/" + KptFileVersion
+)
+
+// KptFile contains information about a package managed with kpt.
+type KptFile struct {
+	yaml.ResourceMeta `yaml:",inline"`
+
+	// Upstream is a reference to where the package is fetched from.
+	Upstream Upstream `yaml:"upstream,omitempty"`
+
+	// PackageMeta contains metadata such as license, documentation, etc.
+	PackageMeta PackageMeta `yaml:"packageMeta,omitempty"`
+
+	// Subpackages declares the list of subpackages.
+	Subpackages []Subpackage `yaml:"subpackages,omitempty"`
+
+	// Pipeline declares the pipeline of functions.
+	Pipeline *Pipeline `yaml:"pipeline,omitempty"`
+
+	// Inventory contains parameters for the inventory object used in apply.
+	Inventory *Inventory `yaml:"inventory,omitempty"`
+}
+
+// OriginType defines the type of origin for a package
+type OriginType string
+
+const (
+	// GitOrigin specifies a package as having been cloned from a git repository
+	GitOrigin OriginType = "git"
+)
+
+// UpdateStrategyType defines the strategy for updating a package from upstream.
+type UpdateStrategyType string
+
+const (
+	// ResourceMerge performs a structural schema-aware comparison and
+	// merges the changes into the local package.
+	ResourceMerge UpdateStrategyType = "resource-merge"
+	// FastForward fails without updating if the local package was modified
+	// since it was fetched.
+	FastForward UpdateStrategyType = "fast-forward"
+	// ForceDeleteReplace wipes all local changes to the package.
+	ForceDeleteReplace UpdateStrategyType = "force-delete-replace"
+)
+
+// Upstream is a reference to where the package is fetched from.
+type Upstream struct {
+	// Type is the type of origin.
+	Type OriginType `yaml:"type,omitempty"`
+
+	// Git contains information on the origin of packages fetched from a git repository.
+	Git Git `yaml:"git,omitempty"`
+
+	// UpdateStrategy defines how a package is updated from upstream.
+	UpdateStrategy UpdateStrategyType `yaml:"updateStrategy,omitempty"`
```

Comment:
What does it mean to have an `UpdateStrategy` as part of `Upstream`? This means that the definition of the subpackage in the parent Kptfile and the upstream information could have different update strategies. It seems like just keeping it in the subpackage section of the parent Kptfile would be sufficient.

Reply:
Just considering the single-package UX for sake of this discussion: Do we want the update strategy of the package to be declared?

1. kpt pkg get foo
2. kpt pkg update

Should step 2 always specify the strategy as a flag? Or should it be recorder in the Kptfile?

Back to the subpackage UX, isn't this a general problem of having Upstream declared in the parent package (`.subpackage.upstream`) and in the subpackage (`.upstream`). You can edit other fields such as Git repo, directory, etc. which needs to be reconciled.

Reply:
So I think there is a couple of things here. 

I agree that being able to specify an update strategy on the root package declaratively in the Kptfile would be useful. It is a bit awkward the way it is now, where it needs to be specified as a flag on every update. We should see if there is a way to address that. It seems like it would require specifying in the package Kptfile, which seems less ideal for subpackages.

Having the git information in both the parent Kptfile and the package Kptfile serves two different purposes. The information in the package Kptfile tells us origin of the current content of the package, which we use for the 3-way merge. The information in the parent Kptfile is where the user specifies the desired version, i.e. the upstream in our 3-way merge. So we actually need both pieces of information, but it not clear to me what it means to have the update strategy in two places.

I think we want to use separate types for subpackage and upstream. Not only to avoid having update strategy in two places, but also to let us keep separate information about the upstream in the two places. Currently we share the `Git` struct between upstream and subpackages, which is a bit unfortunate. The git information specified in subpackages should only require `repo`, `directory` and `ref`, while the information in upstream also needs the `commit` field. We currently allow the `commit` field inside subpackage spec, which we shouldn't.

Reply:
I think we're on the same page for most part. This is what I refer to as "reconciliation". Kpt reconciles the following:

1. Last fetched upstream: `kptfile.upstream`
2. Desired upstream: `kptfile.subpackage.upstream`

Keep in mind, we want to design the schema so it's possible to add other version control systems in the future without breaking backwards compatibility. `kptfile.upstream` has Git in a "one of" which is what we want.

I see couple of options:
1. Use different types for 1 and 2.
2. Reuse common type `Upstream` for both 1 and 2. 

"The git information specified in subpackages should only require repo, directory and ref, while the information in upstream also needs the commit field. We currently allow the commit field inside subpackage spec, which we shouldn't."

It should be fine to specify an exact commit id for the desired origin. All Git-based tooling supports this (Golang, Config Sync, etc.). Providing humna-readable refs should be optional not mandatory. Actually, I realize that `ref` is kind of ambiguous (under-documented here). Does it hold Git tags or branches?

Reply:
I think separating the two types makes sense and gives us the most flexibility going forward. In particular if we want to add support for other storage systems where the information needed for tracking "last fetched upstream" vs "desired upstream" might differ significantly. 

For the `Git` type, the `ref` field holds a git reference, which could be branch, tag or commit. This is the field used by users to specify desired upstream ref. The `commit` field on the other hand, is used to hold a reference to origin for the purpose of keeping track of the last fetched upstream. So even is the user specifies a tag in the `ref` field, kpt will find the commit sha and use that to track origin. We should consider the `commit` field as an implementation detail that is managed by kpt, while `ref` is the api for users (regardless if they want to specify branch, tag or short/long commit sha).

Reply:
Create two types: Upstream and UpstreamLock.

Reply:
@mortent This formulation makes it obvious that we don't have an `Upstream` for the top-level package; only a `UpstreamLock` (We were discussing this in the context of recording update strategy before).

Reply:
Yeah, I think the two different types makes sense. I think we should also create two separate types for the `Git` field in `Upstream` and `UpstreamLock`. The `Commit` field should only be available in `UpstreamLock`. The question around the use of this field on this PR suggests it is confusing.

Reply:
Split `Git` and `GitLock` for maximum decoupling/flexibility.